rfcs/2022-10-12-14742-flexible-metric-tags.md

```diff
@@ -0,0 +1,191 @@
+# RFC 14742 - Support for Duplicate and Bare Tags on Metrics
+
+Vector's current metric model supports only a single value for each tag. The `datadog_agent` source,
+however, may send metrics that contain multiple values for a given tag name, or bare tags with no
+value. Vectors tag support should be enhanced to handle this data.
+
+## Context
```

Comment:
I'm curious if any other sources or sinks allow for tags that don't fit Vector's current model. I poked around a bit and couldn't find any aside from Datadog (and dogstatsd) (which is interesting in itself). Prometheus, AWS CloudWatch, GCP Metrics, New Relic, OpenTelemetry, seem to all use the unique key/value model.

It does mean it seems like we can consider duplicate key names and bare tags as the exception and optimize for the normal key/value case.

Reply:
`OpenTelemetry` attributes support keys having arrays of values which is, more or less, duplicate keys?

Reply:
That's a good point, I neglected OTel allows for the values to be any value including arrays or nested objects. I wonder if that makes them different enough from "tags" where most sinks only allow strings, to put them in the metric event metadata and let users copy them into tags in a `remap` transform if they want (and can handle the different object types).

Reply:
* There is nothing inherent in either of the Prometheus encodings we support that limits them to unique keys or prevents the use of bare tags. The tag value is not explicitly marked as optional in the protocol, but Protobuf allows transmission of data that is missing fields not marked as `required`. Whether or not receivers will accept them is another matter, though. We should accept them.
* AFAICT AWS Cloudwatch metrics can support repeated (they call them "dimensions") as well as optional values. Actually, it looks like they can support an optional tag name as well, which is weird.
* GCP StackDriver/Cloud Monitoring metrics indeed are restricted to a map of key->value pairs, which prevents repeated keys.
* I don't see any evidence the `new_relic` sink makes any use of tags at all.